server/src/utils.ts

```diff
@@ -318,10 +311,53 @@ export let getCompiledFilePath = (
     suffix = bsconfig.suffix;
   }
 
+  return [suffix, pathFragment];
+};
+
+export let getCompiledFilePath = (
+  filePath: string,
+  projDir: string
+): execResult => {
+  let bsconfig = readBsConfig(projDir);
+  let error: execResult = {
+    kind: "error",
+    error: "Could not read bsconfig",
+  };
+
+  if (!bsconfig) {
+    return error;
+  }
+
+  let [suffix, pathFragment] = getSuffixAndPathFragmentFromBsconfig(bsconfig);
   let partialFilePath = filePath.split(projDir)[1];
   let compiledPartialPath = replaceFileExtension(partialFilePath, suffix);
   let result = path.join(projDir, pathFragment, compiledPartialPath);
 
+  // If the file is not found, lookup a possible root bsconfig that may contain
+  // info about the possible location of the file.
+  if (!fs.existsSync(result)) {
+    let rootBsConfigPath = findFilePathFromProjectRoot(
```

Comment:
What exactly should be moved into a separate file? The recursive functions `findProjectRootOfFile` and `findFilePathFromProjectRoot` or the new logic I introduced with this PR (what is inside the `!fs.existsSync` if-block)?

Reply:
I would say just the logic.
I.e. go looking in the parent, find the root from there, take the info from the root bsconfig.

I guess the logic for the binary is similar and if they go to the same place it's easy to keep them in sync.

Reply:
This will introduce circular dependencies. I know this will work in TypeScript, but I think it is not good regardless.
Except I move the lookup methods to the new file as well.

Reply:
Why would move some code to a brand new file introduce circular dependencies?

Unless this refers to something else.z

Reply:
Oh the lookup methods.
Tbh this "utils" file does not really mean very much.
I would not hesitate moving things around and perhaps create new files.

Reply:
And this is definitely not a generic util.

Reply:
Assuming I keep the getCompiledFilePath in utils.ts, but move the new parts into another file. In that file I would have to access `utils.findFilePathFromProjectRoot` among others --> circular dep

Reply:
Looks great. Ready to merge I think.